src/kit/Theme/UI.tsx

```diff
@@ -1,101 +1,93 @@
-  showSpinner: false,
-  theme: {} as Theme,
+const camelCaseToKebab = (text: string): string => {
+  return text
+    .trim()
+    .split('')
+    .map((char, index) => {
+      return char === char.toUpperCase() ? `${index !== 0 ? '-' : ''}${char.toLowerCase()}` : char;
+    })
+    .join('');
 };
 
-const StoreActionUI = {
-  HideUIChrome: 'HideUIChrome',
-  HideUISpinner: 'HideUISpinner',
-  SetMode: 'SetMode',
-  SetPageVisibility: 'SetPageVisibility',
-  SetTheme: 'SetTheme',
-  ShowUIChrome: 'ShowUIChrome',
-  ShowUISpinner: 'ShowUISpinner',
-} as const;
-
-type ActionUI =
-  | { type: typeof StoreActionUI.HideUIChrome }
-  | { type: typeof StoreActionUI.HideUISpinner }
-  | { type: typeof StoreActionUI.SetMode; value: Mode }
-  | { type: typeof StoreActionUI.SetPageVisibility; value: boolean }
-  | { type: typeof StoreActionUI.SetTheme; value: { darkLight: DarkLight; theme: Theme } }
-  | { type: typeof StoreActionUI.ShowUIChrome }
-  | { type: typeof StoreActionUI.ShowUISpinner };
-
-class UIActions {
-  constructor(private dispatch: Dispatch<ActionUI>) {}
-
-  public hideChrome = (): void => {
-    this.dispatch({ type: StoreActionUI.HideUIChrome });
-  };
-
-  public hideSpinner = (): void => {
-    this.dispatch({ type: StoreActionUI.HideUISpinner });
-  };
-
-  public setMode = (mode: Mode): void => {
-    this.dispatch({ type: StoreActionUI.SetMode, value: mode });
-  };
+export const UIProvider: React.FC<{
+  children?: React.ReactNode;
+  themeIsDark?: boolean;
+  theme: Theme;
+}> = ({ children, theme, themeIsDark = false }) => {
+  const ref = useRef(null);
+  return (
+    <UIContext.Provider value={{ ref, theme, themeIsDark }}>
+      <UI theme={theme} themeIsDark={themeIsDark}>
+        <div ref={ref}>{children}</div>
+      </UI>
+    </UIContext.Provider>
+  );
+};
 
-  public setPageVisibility = (isPageHidden: boolean): void => {
-    this.dispatch({ type: StoreActionUI.SetPageVisibility, value: isPageHidden });
-  };
+export const UI: React.FC<{
+  children?: React.ReactNode;
+  themeIsDark?: boolean;
+  theme: Theme;
+}> = ({ children, theme, themeIsDark = false }) => {
+  const ref = useRef<HTMLDivElement>(null);
 
-  public setTheme = (darkLight: DarkLight, theme: Theme): void => {
-    this.dispatch({ type: StoreActionUI.SetTheme, value: { darkLight, theme } });
-  };
+  useLayoutEffect(() => {
+    // Set global CSS variables shared across themes.
+    Object.keys(globalCssVars).forEach((key) => {
+      const value = (globalCssVars as Record<RecordKey, string>)[key];
+      document.documentElement.style.setProperty(`--${camelCaseToKebab(key)}`, value);
+    });
 
-  public showChrome = (): void => {
-    this.dispatch({ type: StoreActionUI.ShowUIChrome });
-  };
+    // Set each theme property as top level CSS variable.
+    Object.keys(theme).forEach((key) => {
+      const value = (theme as Record<RecordKey, string>)[key];
+      ref.current?.style.setProperty(`--theme-${camelCaseToKebab(key)}`, value);
```

Comment:
as mentioned before, we can't rely on inherited styling for the theming because elements can render outside of the context div using portals. Even if we cover everything in the ui kit, that still means userland code can't use portals.

Instead:
* hash the `themeIsDark` and `theme` variables
* create a stylesheet where all elements that have the hash as a class will have the theme variables, and place it in the document head.
* use this class in the context div, as well as any components that render in portals (select, notification, dropdown, modal, drawer, etc.)